include/counter.h

```diff
@@ -102,10 +100,9 @@ struct counter_config_info {
 typedef int (*counter_api_start)(struct device *dev);
 typedef int (*counter_api_stop)(struct device *dev);
 typedef u32_t (*counter_api_read)(struct device *dev);
-typedef int (*counter_api_set_alarm)(struct device *dev,
-				const struct counter_alarm_cfg *alarm_cfg);
-typedef int (*counter_api_disable_alarm)(struct device *dev,
+typedef int (*counter_api_set_alarm)(struct device *dev, u8_t chan_id,
```

Comment:
then driver must implement some kind of allocator. In case of single instance watchdog which is shared between multiple subsystems that's good approach but in case of counter it will most likely be used by single subsystem or object so allocator, imo,  is unnecessary complication of the driver.